### PR TITLE
Issue : upsample_bilinear #7238

### PR DESCRIPTION
**Pull request by tekmen0**

### CODE REVIEW

File: ivy/functional/frontends/torch/nn/functional/vision_functions.py [hunk 1]
```diff
@@ -120,3 +120,108 @@ def pad(input, pad, mode="constant", value=0):
                 + "['constant', 'reflect', 'replicate', 'circular']"
             ).format(mode)
         )
+
+
+@to_ivy_arrays_and_back
+def upsample_bilinear(input, size=None, scale_factor=None):
+    if scale_factor and size:
+        raise ivy.exceptions.IvyException(
```
Comment (nassimberrada) [inline ivy/functional/frontends/torch/nn/functional/vision_functions.py:128]: Given that the function should only take 4-dimensional inputs, another exception could be introduced to handle inputs with dimensions different from 4

File: ivy/functional/frontends/torch/nn/functional/vision_functions.py [hunk 2]
```diff
@@ -120,3 +120,108 @@ def pad(input, pad, mode="constant", value=0):
...
+            ("either size or scale_factor should be defined")
+        )
+
+    n, c, w_old, h_old = input.shape
+
+    if scale_factor and (not size):
+        w_new = int(w_old * scale_factor)
+        h_new = int(h_old * scale_factor)
+    elif (not scale_factor) and size:
+        w_new, h_new = size
```
Comment (nassimberrada) [inline ivy/functional/frontends/torch/nn/functional/vision_functions.py:142]: From my understanding, you are assuming that the size argument is always a tuple, and the scale_factor always an int. From reading the docs for [torch.nn.functional.upsample_bilinear](https://pytorch.org/docs/stable/generated/torch.nn.functional.upsample_bilinear.html) i see that the size argument can also be an int, while the scale_factor can also be a tuple, you may want to take this into consideration

**Comment by nassimberrada** [comment]
Hi @tekmen0, thank you for your contribution !
Your implementation is going in the right direction, with that said I pushed couple of comments which you may want to take a look at.
Also running the following random example indicates that there might be some issues
Using torch:
```
import torch
a = torch.tensor([[[[5., 3.]]]])
torch.nn.functional.upsample_bilinear(a,(5,5))
```
Should result in:
```
tensor([[[[5.0000, 4.5000, 4.0000, 3.5000, 3.0000],
          [5.0000, 4.5000, 4.0000, 3.5000, 3.0000],
          [5.0000, 4.5000, 4.0000, 3.5000, 3.0000],
          [5.0000, 4.5000, 4.0000, 3.5000, 3.0000],
          [5.0000, 4.5000, 4.0000, 3.5000, 3.0000]]]])
```
While using your function with the same parameters, i get this result:
```
IndexError: index 1 is out of bounds for axis 0 with size 1
```
Could you check what may be causing this error ? 
Before doing so, i suggest you merge with the master once again, there was an issue with the CI that prevented torch frontend functions to test properly. The issue should be fixed now so hopefully you can make use of the tests to detect remaining errors.
Let me know if anything is unclear to you, thanks !

**Review by nassimberrada — CHANGES_REQUESTED** [review]

File: ivy/functional/frontends/torch/nn/functional/vision_functions.py (hunk 1 above)
Comment (tekmen0) [reply ivy/functional/frontends/torch/nn/functional/vision_functions.py:146]: Thank you for review. In the original library, this exception is thrown by lower level function torch.nn.functional.interpolate function. Should I still handle even if in the original library it is handled by interpolate (which is not implemented yet in ivy torch frontend)?

File: ivy/functional/frontends/torch/nn/functional/vision_functions.py (hunk 2 above)
Comment (tekmen0) [reply ivy/functional/frontends/torch/nn/functional/vision_functions.py]: Thank you for invaluable review, should be resolved by the latest commits.

File: ivy_tests/test_ivy/test_frontends/test_torch/test_vision_functions.py [hunk 3]
```diff
@@ -171,3 +171,69 @@ def test_torch_pad(
...
+            min_dim_size=2,
+            max_dim_size=1000,
+            min_value=-1e05,
+            max_value=1e05,
+        )
+    )
+    size = None
+    scale_factor = None
+    is_size_used = draw(
+        st.sampled_from(
```
[the remark this answers is not in the document]
Comment (tekmen0) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_vision_functions.py]: Thank you for invaluable review, should be refactored by the latest commits.

**Comment by tekmen0** [comment]
There is an error when upsampling to (n,m) where n is not equal to m.

File: ivy/functional/frontends/torch/nn/functional/vision_functions.py [hunk 4]
```diff
@@ -120,3 +122,78 @@ def pad(input, pad, mode="constant", value=0):
                 + "['constant', 'reflect', 'replicate', 'circular']"
             ).format(mode)
         )
+
+@to_ivy_arrays_and_back
+def upsample_bilinear(input, size=None, scale_factor=None):
```
Comment (tekmen0) [inline ivy/functional/frontends/torch/nn/functional/vision_functions.py:144]: Function works unexpectedly when upsampling input to size = (n,m) where n is not equal to m.

**Comment by nassimberrada** [comment]
Hi @tekmen0, sorry for the wait ! I think your PR is good to merge, the only thing left is the exception handling for different-than-4D inputs. Let me know when that is done and i'll make a final review before merging. Thanks !

**Comment by tekmen0** [comment]
Hi, 4D input handling should be finished : 

[783cbc8](https://github.com/unifyai/ivy/pull/7338/commits/783cbc8fce57536b757a066d4e3a8cbaa672b815)
[9a9513c](https://github.com/unifyai/ivy/pull/7338/commits/9a9513cf6f24fa4b62d326e0e439820f72b04967)

File: ivy/functional/frontends/torch/nn/functional/vision_functions.py (hunk 1 above)
Comment (tekmen0) [reply ivy/functional/frontends/torch/nn/functional/vision_functions.py:145]: Hi, 4D input handling should be finished : 

[783cbc8](https://github.com/unifyai/ivy/pull/7338/commits/783cbc8fce57536b757a066d4e3a8cbaa672b815)
[9a9513c](https://github.com/unifyai/ivy/pull/7338/commits/9a9513cf6f24fa4b62d326e0e439820f72b04967)

**Comment by nassimberrada** [comment]
Looks good to me, i'll be merging then, thank you for your patience !

**Comment by nassimberrada** [comment]
Close #7238

**Pull request merged**